README.md
[![Build Status](https://travis-ci.org/n3okill/enfslist.svg)](https://travis-ci.org/n3okill/enfslist)
[![Build status](https://ci.appveyor.com/api/projects/status/13d4v3xra2f565rt?svg=true)](https://ci.appveyor.com/project/n3okill/enfslist)
[![Codacy Badge](https://api.codacy.com/project/badge/grade/02cba6c651d44e9c96068d3d2d57e89a)](https://www.codacy.com/app/n3okill/enfslist)
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=64PYTCDH5UNZ6)

[![NPM](https://nodei.co/npm/enfslist.png)](https://nodei.co/npm/enfslist/)

enfslist
=========
Module that add list functionality to node fs module

**enfs** stands for [E]asy [N]ode [fs]

This module is intended to work as a sub-module of [enfs](https://www.npmjs.com/package/enfs)

Description
-----------
This module will add a method that allows the obtaining of the 
list of items in the file system under one directory and sub-directories.

- This module will add following methods to node fs module:
  * list
  * listSync
  
Usage
-----
`enfslist`

```js
    var enfslist = require("enfslist");
```

Errors
------
All the methods follows the node culture.
- Async: Every async method returns an Error in the first callback parameter
- Sync: Every sync method throws an Error.


Additional Methods
------------------
- [list](#list)
- [listSync](#listsync)


### list
  - **list(path, [options], callback)**

> Obtain the list of items under a directory and sub-directories asynchronously.
Each item will be an object containing: {path: pathToItem, stat: itemStat}

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * dereference (Boolean): if true will dereference symlinks listing the items to where it points (default: false)
  * back (Number): What you want to get back (0- an array of objects with path and stats, 1- only the paths, 2- only the stats);
  * stats (Array): What information you want in stats. Ex: ```js ["isFile", "atime", "size"] ```
  * ignoreAccessError (Boolean): If true will ignore all files and folder's without access permission (default false)

```js
    enfslist.list("/home", function(err, listOfItems){
        listOfItems.forEach(function(item){
            //do something
        });
    });
```


### listSync
  - **listSync(path, [options])**

> Obtain the list of items under a directory and sub-directories synchronously
Each item will be an object containing: {path: pathToItem, stat: itemStat}

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * dereference (Boolean): if true will dereference symlinks listing the items to where it points (default: false)
  * back (Number): What you want to get back (0- an array of objects with path and stats, 1- only the paths, 2- only the stats);
  * stats (Array): What information you want in stats. Ex: ```js ["isFile", "atime", "size"] ```
  * ignoreAccessError (Boolean): If true will ignore all files and folder's without access permission (default false)

```js
    var listOfItems = enfslist.listSync("/home");
    listOfItems.forEach(function(item){
        //do something
    });
```


License
-------

Creative Commons Attribution 4.0 International License

Copyright (c) 2016 Joao Parreira <joaofrparreira@gmail.com> [GitHub](https://github.com/n3okill)

This work is licensed under the Creative Commons Attribution 4.0 International License. 
To view a copy of this license, visit [CC-BY-4.0](http://creativecommons.org/licenses/by/4.0/).




CHANGELOG.md
1.0.0 / 2017-03-31
==================
  * code improvement
  * dependencies update
  * major version

0.1.0 / 2016-04-06
==================
  * es6 code update
  * add back and stats to options
  * dependency packages upgrade
  * added ignoreAccessError to options

0.0.1 / 2016-02-20
==================

  * first release


test/util.js
/**
 * @project enf
 * @filename util.js
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2017 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 15-02-2017
 * @version 0.0.1
 * @description
 */

"use strict";

const fs = require("fs");
const rimraf = require("rimraf");
const cwd = process.cwd();

module.exports = {
    before: (path, files, done) => {
        fs.mkdir(path, (err) => {
            (err === null).should.be.equal(true);
            process.chdir(path);
            let filesLength = files.length;
            files.forEach((p) => {
                fs.writeFile(p, "data", "utf8", (errWrite) => {
                    (errWrite === null).should.be.equal(true);
                    if (--filesLength === 0) {
                        done();
                    }
                });
            });
        });
    },
    after: (path, done) => {
        process.chdir(cwd);
        rimraf(path, done);
    }
};

test/listSync.js
/**
 * @project enfslist
 * @filename listSync.js
 * @description async methods for listing items in directories
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.1
 */
/* global describe, it, before, after*/
"use strict";

const nodePath = require("path");
const nodeOs = require("os");
const enfsList = require("../");
const util = require("./util");

describe("enfslist sync", function () {
    const tmpPath = nodePath.join(nodeOs.tmpdir(), "enfslistsync");
    const files = ["file1", "file2", "file3", "file4"];
    before(function (done) {
        util.before(tmpPath, files, done);
    });
    after(function (done) {
        util.after(tmpPath, done);
    });
    it("should list files", function () {
        //it will return also the base path used to list files
        (function () {
            let list = enfsList.listSync(tmpPath);
            list.length.should.be.equal(files.length + 1); //the +1 is the directory that also appears in the list
        }).should.not.throw();
    });
});


test/listAsync.js
/**
 * @project enfslist
 * @filename listAsync.js
 * @description async methods for listing items in directories
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.1
 */

/* global describe, it, before, after*/
"use strict";

const nodePath = require("path");
const nodeOs = require("os");
const enfsList = require("../");
const util = require("./util");

describe("enfslist async", function () {
    const tmpPath = nodePath.join(nodeOs.tmpdir(), "enfslist");
    const files = ["file1", "file2", "file3", "file4"];
    before(function (done) {
        util.before(tmpPath, files, done);
    });
    after(function (done) {
        util.after(tmpPath, done);
    });
    it("should list files", function (done) {
        //it will return also the base path used to list files
        enfsList.list(tmpPath, function (err, list) {
            (err === null).should.be.equal(true);
            list.length.should.be.equal(files.length + 1); //the +1 is the directory that also appears in the list
            done();
        });
    });
});


lib/util.js
/**
 * @project enf
 * @filename util.js
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2017 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 15-02-2017
 * @version 0.1.1
 * @description utilities for enfslist
 */

"use strict";

const enFs = require("enfspatch");

function getOptions(opt) {
    let options = opt || {};
    options.fs = options.fs || enFs;
    options.dereference = options.dereference === true;
    options.back = options.back || 0;
    options.ignoreAccessError = "ignoreAccessError" in options ? options.ignoreAccessError : false;
    let stats2 = new options.fs.Stats();
    options.showStats = Array.isArray(options.stats) ? options.stats : Object.getOwnPropertyNames(stats2).concat(Object.getOwnPropertyNames(Object.getPrototypeOf(stats2)));
    return options;
}

function getResult(options, path, stats) {
    let miniStats = {};
    options.showStats.forEach((k) => miniStats[k] = stats[k]);
    return options.back === 2 ? miniStats : options.back === 1 ? path : {path, stat: miniStats};
}

module.exports = {
    getOptions,
    getResult
};

lib/listSync.js
/**
 * @project enfslist
 * @filename listSync.js
 * @description sync methods for listing items in directories
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.1.1
 */


"use strict";

const nodePath = require("path");
const nodeAssert = require("assert");
const util = require("./util");

/**
 * List all items inside a directory and sub-directories
 * @param {string} path - the path to the directory
 * @param {object} opt - various options for list module
 *              {bool} opt.dereference - defines the type of stat that will be used in files
 *              {object} opt.fs - the fs module to be used
 *              {number} opt.back - the elements to return (0- both[path and stats], 1 - path, 2 - stats)
 *              {array} opt.stats - return only the required stats information
 *              {bool} opt.ignoreAccessError - if true will ignore all directories without reading access (default false)
 * @return {Error|Array}
 */

function list(path, opt) {
    let items = [], result = [];
    nodeAssert(path, "path must be defined");

    const options = util.getOptions(opt);
    options.stat = options.dereference ? options.fs.statSync : options.fs.lstatSync;

    items.push(path);
    do {
        const item = items.shift();
        let stats;
        try {
            stats = options.stat(item);
        } catch (err) {
            if (!(err.code === "ENOENT" && options.ignoreAccessError)) {
                throw err;
            }
        }
        result.push(util.getResult(options,item,stats));
        if (stats.isDirectory()) {
            try {
                items = items.concat(options.fs.readdirSync(item).map((m) => {
                    return nodePath.join(item, m);
                }));
            } catch (err) {
                if (!options.ignoreAccessError || (options.ignoreAccessError && !(err.code === "EACCES" || err.code === "EPERM"))) {
                    throw err;
                }
            }
        }
    } while (items.length);

    return result;
}


module.exports = list;


lib/listAsync.js
/**
 * @project enfslist
 * @filename listAsync.js
 * @description async methods for listing items in directories
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.1.0
 */

"use strict";


const nodePath = require("path");
const nodeAssert = require("assert");
const util = require("./util");




function listFiles(path, options, callback) {
    let result = [];
    options.stat(path, (err, stats) => {
        if (err) {
            if (err.code === "ENOENT" && options.ignoreAccessError) {
                return callback(null, result);
            }
            return callback(err);
        }
        result.push(util.getResult(options, path, stats));
        if (stats.isDirectory()) {
            options.fs.readdir(path, (errReadDir, files) => {
                if (errReadDir) {
                    if (!options.ignoreAccessError || (options.ignoreAccessError && !(errReadDir.code === "EACCES" || errReadDir.code === "EPERM"))) {
                        return callback(errReadDir);
                    }
                    return callback(null, result);
                }
                let filesLength = files.length;
                if (filesLength === 0) {
                    return callback(null, result);
                }
                files.forEach((file) => {
                    let filePath = nodePath.join(path, file);
                    listFiles(filePath, options, (errList, r) => {
                        if (errList) {
                            return callback(errList);
                        }
                        result = result.concat(r);
                        if (--filesLength === 0) {
                            return callback(null, result);
                        }
                    });
                });
            });
        } else {
            return callback(null, result);
        }
    });
}


/**
 * List all items inside a directory and sub-directories
 * @param {string} path - the path to the directory
 * @param {object} opt - various options for list module
 *              {bool} opt.dereference - defines the type of stat that will be used in files
 *              {object} opt.fs - the fs module to be used
 *              {number} opt.back - the elements to return (0- both[path and stats], 1 - path, 2 - stats)
 *              {array} opt.stats - return only the required stats information
 *              {bool} opt.ignoreAccessError - if true will ignore all files that can't be accessed (default false)
 * @param {function} callback - the callback function that will be called after the list is done
 * @return {Error|Array}
 */
function list(path, opt, callback) {
    if (typeof opt === "function" && !callback) {
        callback = opt;
        opt = {};
    }

    nodeAssert(path, "'path' is required");
    nodeAssert(callback, "'callback' is required");

    const options = util.getOptions(opt);
    options.stat = options.dereference ? options.fs.stat : options.fs.lstat;

    listFiles(path, options, callback);
}

module.exports = list;